Create default SCC for cluster

pweil- opened the pull request:
Right now just a run as any SCC until we solve the issue with volume permissions and identifying build pod creators.  Will follow up with a real set of SCCs.

Based on https://github.com/openshift/origin/pull/2856 which should be merged first.  Only the second commit is new code.

Fixes https://github.com/openshift/origin/issues/2775

@liggitt @pmorie @smarterclayton

pweil- commented:
@derekwaynecarr @deads2k - tagging for more reviewers

pweil- commented:
rebased

smarterclayton commented on pkg/cmd/server/start/start_master.go:
```diff
@@ -411,6 +411,7 @@ func StartMaster(openshiftMasterConfig *configapi.MasterConfig) error {
 			openshiftConfig.RunDeploymentImageChangeTriggerController()
 			openshiftConfig.RunImageImportController()
 			openshiftConfig.RunOriginNamespaceController()
+			openshiftConfig.BootstrapSecurityContextConstraints()
```
This should be called with the other bootstrappers - it's not a controller.

liggitt replied:
Paul, see ensureComponentAuthorizationRules, etc

smarterclayton commented on pkg/cmd/server/origin/master.go:
```diff
@@ -1063,6 +1065,35 @@ func (c *MasterConfig) RunSecurityAllocationController() {
+		ObjectMeta: kapi.ObjectMeta{
+			Name: "scc-default",
+		},
+		AllowPrivilegedContainer: true,
```
The default scc name should be configurable.  The default one we create should relate to the default security mode we want.  We need to default to a secured model (no privileged, no host dir).

pweil- replied:
10-4.  This is just to get the mechanics right.  The secured model will create SCCs that are very close to https://github.com/openshift/origin/pull/2609#issuecomment-108608800 once the dependent PRs land.

liggitt replied:
Nothing selects a default constraint by name, a constraint becomes the default by virtue of having all users/groups mapped to it

smarterclayton replied:
Let's have the discussion on this in the other thread.

smarterclayton commented:
Move the bootstrapper to be where the policy bootstrapper is and I'll merge this

liggitt commented:
Are we following the same approach as policy? Will create it if it's missing, but leave it alone if it already exists?

pweil- commented:
Updated to move to the ensure... section.  Also added the user scc as @liggitt, we can update who it is granted to when the other PRs land.

@liggitt - yes, it first looks for the presence of any SCCs, if any are found it does nothing with the assumption that the admin is in control.

```
[vagrant@openshiftdev origin]$ osc get scc
NAME        PRIV      CAPS      HOSTDIR   SELINUX     RUNASUSER
scc-admin   true      []        true      RunAsAny    RunAsAny
scc-user    false     []        false     MustRunAs   MustRunAsRange
[vagrant@openshiftdev origin]$ osc get scc scc-admin -o json
{
    "kind": "SecurityContextConstraints",
    "apiVersion": "v1beta3",
    "metadata": {
        "name": "scc-admin",
        "selfLink": "/api/v1beta3/securitycontextconstraints/scc-admin",
        "uid": "dfc0cd6c-0e04-11e5-b107-080027c5bfa9",
        "resourceVersion": "55",
        "creationTimestamp": "2015-06-08T17:36:20Z"
    },
    "allowPrivilegedContainer": true,
    "allowHostDirVolumePlugin": true,
    "seLinuxContext": {
        "type": "RunAsAny"
    },
    "runAsUser": {
        "type": "RunAsAny"
    },
    "groups": [
        "system:authenticated"
    ]
}
[vagrant@openshiftdev origin]$ osc get scc scc-user -o json
{
    "kind": "SecurityContextConstraints",
    "apiVersion": "v1beta3",
    "metadata": {
        "name": "scc-user",
        "selfLink": "/api/v1beta3/securitycontextconstraints/scc-user",
        "uid": "dfc12d5f-0e04-11e5-b107-080027c5bfa9",
        "resourceVersion": "56",
        "creationTimestamp": "2015-06-08T17:36:20Z"
    },
    "seLinuxContext": {
        "type": "MustRunAs"
    },
    "runAsUser": {
        "type": "MustRunAsRange"
    }
}
```

pweil- commented:
[test]

liggitt commented:
can we remove omitempty for allowPrivilegedContainer and allowHostDirVolumePlugin? It gives me warm fuzzier to see those explicitly denied in the user SCC

liggitt commented:
Why prefix names with "scc-"? They're already typed

liggitt commented:
I'd prefer names like privileged and restricted

pweil- commented:
> can we remove omitempty for allowPrivilegedContainer and allowHostDirVolumePlugin? It gives me warm fuzzier to see those explicitly denied in the user SCC

Sure, I'll add it to the follow up issue since it's mainly just a feel good.


> I'd prefer names like privileged and restricted
I can change all these names when I resubmit to upstream


> Why prefix names with "scc-"? They're already typed

updated

liggitt commented:
I actually meant for names of the default sccs:

scc-admin -> privileged
scc-user -> restricted

liggitt commented on pkg/cmd/server/origin/master.go:
```diff
@@ -1063,6 +1067,51 @@ func (c *MasterConfig) RunSecurityAllocationController() {
+			Name: "user",
+		},
+		SELinuxContext: kapi.SELinuxContextStrategyOptions{
+			Type: kapi.SELinuxStrategyMustRunAs,
```
Add a comment that this requires annotations on the namespace and will reject all pods if those annotations aren't set

smarterclayton replied:
Is the rejection error from the admission controller clear about what is happening?  "namespace does not have an allocated UID or MCS label" etc

pweil- replied:
The admission controller will issue this error if it is unable to find the annotation and will fail

```
annotationVal, ok := ns.Annotations[allocator.UIDRangeAnnotation]
	if !ok {
		return nil, nil, errors.NewInternalError(fmt.Errorf("unable to find annotation %s", allocator.UIDRangeAnnotation))
	}
```

smarterclayton replied:
Not good enough.

----- Original Message -----
> > +		RunAsUser: kapi.RunAsUserStrategyOptions{
> > +			Type: kapi.RunAsUserStrategyRunAsAny,
> > +		},
> > +		Groups: []string{"system:authenticated"},
> > +	}
> > +	_, err = c.KubeClient().SecurityContextConstraints().Create(sccAdmin)
> > +	if err != nil {
> > +		glog.Fatalf("Unable to create default security context constraint %#v.
> > Got error: %v", sccAdmin, err)
> > +	}
> > +
> > +	sccUser := &kapi.SecurityContextConstraints{
> > +		ObjectMeta: kapi.ObjectMeta{
> > +			Name: "user",
> > +		},
> > +		SELinuxContext: kapi.SELinuxContextStrategyOptions{
> > +			Type: kapi.SELinuxStrategyMustRunAs,
> 
> The admission controller will issue this error if it is unable to find the
> annotation and will fail
> 
> ```
> annotationVal, ok := ns.Annotations[allocator.UIDRangeAnnotation]
> 	if !ok {
> 		return nil, nil, errors.NewInternalError(fmt.Errorf("unable to find
> 		annotation %s", allocator.UIDRangeAnnotation))
> 	}
> ```
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2863/files#r31941932

pweil- replied:
Ok.  I'll need to rebase admission after this lands anyway.  Will update to something like "Strategy %s for security context constraint %s requires pre-allocated annotation %s on namespace %s"

liggitt commented on pkg/cmd/server/origin/master.go:
```diff
@@ -1063,6 +1067,51 @@ func (c *MasterConfig) RunSecurityAllocationController() {
+			Type: kapi.SELinuxStrategyMustRunAs,
+		},
+		RunAsUser: kapi.RunAsUserStrategyOptions{
+			Type: kapi.RunAsUserStrategyMustRunAsRange,
```
Same comment here

liggitt replied:
Are these the most restrictive policies that exist?

pweil- replied:
Yes for the SELinux strategy since it is an exact match, no for the user strategy.  The most restrictive user strategy would be `MustRunAs` with a single UID.

smarterclayton commented:
Yeah nuke the prefixes.

pweil- commented:
names updated, comment added.

liggitt commented on pkg/cmd/server/origin/master.go:
```diff
@@ -1063,6 +1067,57 @@ func (c *MasterConfig) RunSecurityAllocationController() {
+		RunAsUser: kapi.RunAsUserStrategyOptions{
+			Type: kapi.RunAsUserStrategyRunAsAny,
+		},
+		Groups: []string{"system:authenticated"},
```
Use the constant defined for this in bootstrappolicy

liggitt replied:
And cluster admins

liggitt replied:
And a todo to remove allauthenticated from this list asap

liggitt commented:
Nits on constants/comments, then LGTM

pweil- commented:
Updated

```
[vagrant@openshiftdev origin]$ osc get scc
NAME         PRIV      CAPS      HOSTDIR   SELINUX     RUNASUSER
privileged   true      []        true      RunAsAny    RunAsAny
restricted   false     []        false     MustRunAs   MustRunAsRange
[vagrant@openshiftdev origin]$ osc get scc privileged -o json
{
    "kind": "SecurityContextConstraints",
    "apiVersion": "v1beta3",
    "metadata": {
        "name": "privileged",
        "selfLink": "/api/v1beta3/securitycontextconstraints/privileged",
        "uid": "0b2814f6-0e0a-11e5-acb2-080027c5bfa9",
        "resourceVersion": "55",
        "creationTimestamp": "2015-06-08T18:13:20Z"
    },
    "allowPrivilegedContainer": true,
    "allowHostDirVolumePlugin": true,
    "seLinuxContext": {
        "type": "RunAsAny"
    },
    "runAsUser": {
        "type": "RunAsAny"
    },
    "groups": [
        "system:authenticated",
        "system:cluster-admins"
    ]
}
[vagrant@openshiftdev origin]$ osc get scc restricted -o json
{
    "kind": "SecurityContextConstraints",
    "apiVersion": "v1beta3",
    "metadata": {
        "name": "restricted",
        "selfLink": "/api/v1beta3/securitycontextconstraints/restricted",
        "uid": "0b2882eb-0e0a-11e5-acb2-080027c5bfa9",
        "resourceVersion": "56",
        "creationTimestamp": "2015-06-08T18:13:20Z"
    },
    "seLinuxContext": {
        "type": "MustRunAs"
    },
    "runAsUser": {
        "type": "MustRunAsRange"
    }
}
```

liggitt commented on pkg/cmd/server/origin/master.go:
```diff
@@ -1063,6 +1067,58 @@ func (c *MasterConfig) RunSecurityAllocationController() {
+		return
+	}
+	glog.Infof("No security context constraints detected, adding defaults")
+	sccAdmin := &kapi.SecurityContextConstraints{
```
define the bootstrap SCC's, then iterate over them and do the create. errors probably shouldn't terminate the server, just log as errors, since the cluster admin can create them in a running server

liggitt commented:
other than the iterate/log comment, I'm fine with this for now. @smarterclayton ?

pmorie commented on pkg/cmd/server/origin/master.go:
```diff
@@ -1063,6 +1067,58 @@ func (c *MasterConfig) RunSecurityAllocationController() {
+
+	sccUser := &kapi.SecurityContextConstraints{
+		ObjectMeta: kapi.ObjectMeta{
+			Name: "restricted",
```
Make this and `privileged` constants.

liggitt commented:
hmm... since we don't overwrite existing, anyone who runs once this goes in will have an SCC that allows anyone root access... we have a plan for making sure no one gets bitten by that?

smarterclayton commented:
Approved, no issues, you own the merge

> On Jun 8, 2015, at 6:07 PM, Jordan Liggitt <notifications@github.com> wrote:
> 
> other than the iterate/log comment, I'm fine with this for now. @smarterclayton ?
> 
> —
> Reply to this email directly or view it on GitHub.
>

pweil- commented:
updated to use a slice of SCs and iterate, replace Fatalf with Errorf, added constants for names.  

> hmm... since we don't overwrite existing, anyone who runs once this goes in will have an SCC that allows anyone root access... we have a plan for making sure no one gets bitten by that?

After startup it becomes an admin's job to manage this.  We could make it easier with a recreate command to recreate defaults, it should be admin triggered action.

smarterclayton commented:
Or they delete them an restart their server

> On Jun 8, 2015, at 8:36 PM, Paul Weil <notifications@github.com> wrote:
> 
> updated to use a slice of SCs and iterate, replace Fatalf with Errorf, added constants for names.
> 
> hmm... since we don't overwrite existing, anyone who runs once this goes in will have an SCC that allows anyone root access... we have a plan for making sure no one gets bitten by that?
> 
> After startup it becomes an admin's job to manage this. We could make it easier with a recreate command to recreate defaults, it should be admin triggered action.
> 
> —
> Reply to this email directly or view it on GitHub.
>

liggitt commented:
fair enough... it only affects pod creation, so it's less dicey than stomping policy, which can actually prevent you from fixing it via the API

liggitt commented:
LGTM, [merge]

liggitt commented:
squash, then you get the merge

pweil- commented:
lulz....I must have copy/pasted the wrong hash for my reset.  Done.

liggitt commented on pkg/cmd/server/origin/master.go:
```diff
@@ -1063,6 +1067,23 @@ func (c *MasterConfig) RunSecurityAllocationController() {
 	controller.Run()
 }
 
+func (c *MasterConfig) ensureDefaultSecurityContextConstraints() {
```
just realized where this function was... move it up with the other ensure* functions

pweil- commented:
:musical_note: movin on up...done

liggitt commented:
[merge]

pweil- commented:
```
!!! Error in hack/test-cmd.sh:616
  'oc delete all --all' exited with status 1
```

liggitt commented:
only the merge queue matters now!
